### PR TITLE
[perf, fsdp, trainer] feat: Skip training for zero-advantage responses to speed up RL.

### PR DESCRIPTION
**Pull request by sheilaliuxl**
### What does this PR do?

**SKIP** training for zero-advantage responses to speed up RL
- On top of #5770 

At high accuracy, less than `25%` samples have non-zero advantage (`critic/advantages/zero_adv_ratio`), and they don't contribute to the `pg_loss` at all. With `4` mini-batches:
1. It can train with the same `#mini-batches` as before, but with less `#samples`: `match_loss_curve = True`
    - This option is designed to match baseline as much as possible, including `num_ghost_opt_steps` opt steps when needed (Very unlikely)
2. It can be reduced into `1` single mini-batch: `match_loss_curve = False` 
    - This option is designed to train as fast as possible with the same `micro-bs` (static bs) or `tokens_per_gpu` (dynamic bs), faster than both baseline and `#1` option

Both options closely match baseline loss curve, and train faster due to the `skip`.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+zero+adv+is%3Aopen
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`

#### 1. `Qwen2.5 3B LoRA`: `2` epochs for `GSM8k`

Metrics comparison dropping noisy first `10` step and last `3` steps:
1. `timing_s/update_actor`
    - `avg`: `3.853 -> 2.278 (-40.9%)` or `1.7x` reduction
    - `std`: `0.183 -> 0.072 (-60.7%)` or `2.5x` reduction
2. `critic/rewards/mean`
    - `avg`: `0.909 -> 0.915`
8. `val-core/openai/gsm8k/acc/mean@1` 
    - `avg`: `0.855 -> 0.861`

```
sliuxl@c889f3b957d9 20:35 ~/qwen2d5 $ M=timing_s/update_actor; conda run -n et python ~/tb.py --files "sliuxl-mverl--qwen2d5_3b_inst--gsm8k/*/events*" --metric "$M"
2026-03-31 20:35:42,832 [tb.py:206] WARNING - Unknonw metric `timing_s/update_actor`!
2026-03-31 20:35:43,069 [tb.py:235] INFO - Set benchmark with index 0
2026-03-31 20:35:43,070 [tb.py:264] INFO - [103/116] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13954-run-mverl-rl-0adv-05-j-01-ns01-tp01-bs16-pad0-base--20260331.152146/events.out.tfevents.1774996037.ip-10-4-145-27.2208577.0: (avg (std), med) = (  3.853 (  0.183),   3.814)  # timing_s/update_actor
2026-03-31 20:35:43,337 [tb.py:264] INFO - [103/116] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13955-run-mverl-rl-0adv-05-j-02-ns01-tp01-bs16-pad0-rm-0adv-yes-mlc--20260331.152151/events.out.tfevents.1775000615.ip-10-4-145-27.2332233.0: (avg (std), med) = (  2.278 (  0.072),   2.254)  # timing_s/update_actor
2026-03-31 20:35:43,606 [tb.py:264] INFO - [103/116] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13956-run-mverl-rl-0adv-05-j-03-ns01-tp01-bs16-pad0-rm-0adv-no-mlc--20260331.152156/events.out.tfevents.1775004996.ip-10-4-145-27.2451478.0: (avg (std), med) = (  1.652 (  0.473),   1.527)  # timing_s/update_actor

sliuxl@c889f3b957d9 19:56 ~/qwen2d5 $ M=critic/rewards/mean; conda run -n et python ~/tb.py --files "sliuxl-mverl--qwen2d5_3b_inst--gsm8k/*/events*" --metric "$M"
2026-03-31 19:57:02,837 [tb.py:206] WARNING - Unknonw metric `critic/rewards/mean`!
2026-03-31 19:57:03,085 [tb.py:235] INFO - Set benchmark with index 0
2026-03-31 19:57:03,086 [tb.py:264] INFO - [103/116] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13954-run-mverl-rl-0adv-05-j-01-ns01-tp01-bs16-pad0-base--20260331.152146/events.out.tfevents.1774996037.ip-10-4-145-27.2208577.0: (avg (std), med) = (  0.909 (  0.028),   0.913)  # critic/rewards/mean
2026-03-31 19:57:03,348 [tb.py:264] INFO - [103/116] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13955-run-mverl-rl-0adv-05-j-02-ns01-tp01-bs16-pad0-rm-0adv-yes-mlc--20260331.152151/events.out.tfevents.1775000615.ip-10-4-145-27.2332233.0: (avg (std), med) = (  0.915 (  0.026),   0.918)  # critic/rewards/mean
2026-03-31 19:57:03,630 [tb.py:264] INFO - [103/116] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13956-run-mverl-rl-0adv-05-j-03-ns01-tp01-bs16-pad0-rm-0adv-no-mlc--20260331.152156/events.out.tfevents.1775004996.ip-10-4-145-27.2451478.0: (avg (std), med) = (  0.904 (  0.027),   0.906)  # critic/rewards/mean


sliuxl@c889f3b957d9 20:35 ~/qwen2d5 $ M=val-core/openai/gsm8k/acc/mean@1; conda run -n et python ~/tb.py --files "sliuxl-mverl--qwen2d5_3b_inst--gsm8k/*/events*" --metric "$M"
2026-03-31 20:38:39,184 [tb.py:206] WARNING - Unknonw metric `val-core/openai/gsm8k/acc/mean@1`!
2026-03-31 20:38:39,432 [tb.py:235] INFO - Set benchmark with index 0
2026-03-31 20:38:39,434 [tb.py:264] INFO - [12/25] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13954-run-mverl-rl-0adv-05-j-01-ns01-tp01-bs16-pad0-base--20260331.152146/events.out.tfevents.1774996037.ip-10-4-145-27.2208577.0: (avg (std), med) = (  0.855 (  0.004),   0.857)  # val-core/openai/gsm8k/acc/mean@1
2026-03-31 20:38:39,695 [tb.py:264] INFO - [12/25] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13955-run-mverl-rl-0adv-05-j-02-ns01-tp01-bs16-pad0-rm-0adv-yes-mlc--20260331.152151/events.out.tfevents.1775000615.ip-10-4-145-27.2332233.0: (avg (std), med) = (  0.861 (  0.005),   0.863)  # val-core/openai/gsm8k/acc/mean@1
2026-03-31 20:38:39,972 [tb.py:264] INFO - [12/25] sliuxl-mverl--qwen2d5_3b_inst--gsm8k/verl_feature_config_pbtxt-dc7fdce1--13956-run-mverl-rl-0adv-05-j-03-ns01-tp01-bs16-pad0-rm-0adv-no-mlc--20260331.152156/events.out.tfevents.1775004996.ip-10-4-145-27.2451478.0: (avg (std), med) = (  0.859 (  0.004),   0.860)  # val-core/openai/gsm8k/acc/mean@1
```
 

Screenshot 2026-03-31 at 19 53 42

### Match Loss Curve

Based on loss agg mode:
1. `token-mean`: `tm`
2. `sequence-mean-token-mean`: `smtm`
3. `sequence-mean-token-sum`: `smts`
5. `sequence-mean-token-sum-norm`: `smtsn`
    - `0.5B`: **Slower** than baseline
    - `3B`: Matches with baseline

Notation: `#mini-batches` is `#optimizer-steps`
1. `mlcTmini`: `(match_loss_curve, match_data_split_mini_batch) = (True, True)`
   * Mathematically should be **equivalent to** baseline i.e. **LOSSLESS**: Preserve mini batch *and* micro-batch (grad accumulation) **data** split, and skip `0` adv samples in each grad accumulation on a GPU (Keep the shortest one when all `0` adv)
2. `mlcT`: `(match_loss_curve, match_data_split_mini_batch) = (True, False)`
   * Preserves `#mini-batches` (not data), but not `#micro-batches`: Slightly more efficient than `#1`
4. `mlcF`: `match_loss_curve = False`
   * It doesn't even try to preserve `#mini-batches`: Most efficient among the `3` options

#### 1. `(0.5B, tm)`
Screenshot 2026-04-09 at 17 14 10--0 5B--tm

#### 2. `(0.5B, smtm)`

Screenshot 2026-04-09 at 17 12 26--0 5B--smtm

#### 3. `(0.5B, smts)`
Screenshot 2026-04-09 at 16 59 47--0 5B--smts-2x

#### 4. `(0.5B, smtsn)`: Slower than baseline

Screenshot 2026-04-09 at 17 54 51--0 5B--smtsn--slower-than-baseline

#### 4.1 `(3B, smtsn)`
Screenshot 2026-04-09 at 18 31 51--03B--smtsn

### API and Usage Example

Add to your RL config with `FSDP`:

```
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.use_kl_loss=False \

    algorithm.filter_zero_adv.enable=True \
    algorithm.filter_zero_adv.match_loss_curve=True \
    algorithm.filter_zero_adv.match_mini_batch_data_split=True \

    trainer.use_legacy_worker_impl=auto \
```

**Known Limitation**
- Strict: To match baseline loss curve, it requires turning off both `entropy_loss` and `kl_loss`, currently hard coded to raise an `Exception` if it violates the condition
- When `kl_loss_coef` is sufficiently small, it's up to the user to turn on this feature: No longer lossless, while a good approximation by keeping KL divergence for `!0` adv samples

### Design & Code Changes

`dp_actor.py` has `3` `loss` components, when both `{#2, #3}` are off, we can skip `loss` completely as the `pg_loss` has `advantage` as a weight factor:
1. `policy_loss` $\propto$ `adv`
9. `entropy_loss`: By `(calculate_entropy: bool, entropy_coeff: float)`  and
10. `kl_loss`: By `(use_kl_loss: bool, kl_loss_coeff: float)`

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/workers/actor/dp_actor.py [hunk 1]
```diff
@@ -565,14 +619,21 @@ def update_policy(self, data: DataProto):
...
+                    if match_loss_curve:
+                        # Use config GA so each sample weighs 1/ppo_mini_batch_size,
+                        # matching baseline per-sample gradient.
+                        self.gradient_accumulation = (
+                            self.config.ppo_mini_batch_size // self.config.ppo_micro_batch_size_per_gpu
+                        )
+                    else:
+                        # Use actual micro-batch count: naturally handles partial last
+                        # mini-batch in the fewer-K path without extra correction.
+                        self.gradient_accumulation = len(micro_batches)
```
[the remark this answers is not in the document]
Comment (sheilaliuxl) [reply verl/workers/actor/dp_actor.py:700]: This is actually intentional for the `match_loss_curve = False` option
- Useful when we extend this feature to the `gen` step later, without keeping the `0 adv` samples at all

File: verl/workers/actor/dp_actor.py [hunk 2]
```diff
@@ -678,5 +755,15 @@ def update_policy(self, data: DataProto):
...
                 append_to_dict(metrics, mini_batch_metrics)
+
+            # Ghost optimizer.step() with zero gradients to maintain K
+            # (the original number of optimizer steps per epoch).
+            # With match_loss_curve's even distribution, this is typically 0.
+            if match_loss_curve:
+                for _ in range(max(0, num_ghost_opt_steps)):
+                    self.actor_optimizer.zero_grad()
+                    grad_norm = self._optimizer_step()
+                    append_to_dict(metrics, {"actor/grad_norm": grad_norm.detach().item()})
```
[the remark this answers is not in the document]
Comment (sheilaliuxl) [reply verl/workers/actor/dp_actor.py:867]: This is desired for `match_loss_curve = True`:
- It's actually **very rare** here, when and only when `#non-zero-adv-samples < #original-opt-steps`
